### PR TITLE
Pinot Spark Connector for Spark3

### PR DESCRIPTION
**Pull request by cbalci**
**Background**
Apache Spark has [changed](https://blog.madhukaraphatak.com/spark-3-datasource-v2-part-3) the Datasource interface significantly between Spark2 and Spark3, so current pinot-spark-connector doesn't work for Spark3. In a previous PR(#10321) I refactored the spark-connector into two modules (`pinot-spark-common` and `pinot-spark-2-connector`) to be able to reuse shared logic which gives us a clean base to implement the new version.

**Change**
In this PR I'm implementing the DataSourceV2 interface as published by Spark3. Functionality is exactly same as Pinot Spark 2 Connector and it supports all existing configuration options such as:
- Ability to read from REALTIME, OFFLINE or HYBRID Pinot tables
- Ability to scan using HTTP or GRPC server endpoints
- Column pruning and filter push down
- etc. (see docs)

It can be used as a drop in replacement when migrating from Spark2 to Spark3. Spark3 also brings some new features and improvements such as 'Aggregation push down' which can be taken advantage of in the future.

**Testing**
I added basic unit test coverage as well as a good list of integration tests under `ExampleSparkPinotConnectorTest` similar to Spark2 Connector. 

`feature`
`release-notes` (Added Spark3 support for Pinot Spark Connector)

### CODE REVIEW

File: pinot-connectors/pinot-spark-3-connector/pom.xml [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+                  <goal>shade</goal>
+                </goals>
+                <configuration>
+                  <relocations>
+                    <relocation>
+                      <pattern>com</pattern>
+                      <shadedPattern>${shadeBase}.com</shadedPattern>
+                      <includes>
+                        <include>com.google.protobuf.**</include>
+                        <include>com.google.common.**</include>
```
Comment (KKcorps) [inline pinot-connectors/pinot-spark-3-connector/pom.xml:153]: Can we check for hadoop classes and also shade them as well.

**Comment by KKcorps** [comment]
Thanks for the contribution! Can you try to test it in cluster mode while running on a proper YARN, AWS EMR or DataProc cluster.

Some of the times spark jobs inside pinot fail in these environments because of some conflicting runtime libs. 

Solution is simply to keep on shading until the problems get resolved imo.

File: pinot-connectors/pinot-spark-3-connector/documentation/read_model.md [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+| partition1  | realtimeServer1 / segment1 |
+| partition2  | realtimeServer1 / segment2  |
+| partition3  | realtimeServer1 / segment3 |
+| partition4  | realtimeServer2 / segment4 |
+| partition5  | offlineServer10 / segment10 |
+| partition6  | offlineServer10 / segment20 |
+
+If `segmentsPerSplit` value is too low, that means more parallelism. But this also mean that a lot of connection will be opened with Pinot servers, and will increase QPS on the Pinot servers. 
+
+If `segmetnsPerSplit` value is too high, that means less parallelism. Each Pinot server will scan more segments per request.
```
Comment (GSharayu) [inline pinot-connectors/pinot-spark-3-connector/documentation/read_model.md:77]: (nit) typo segmentsPerSplit

File: pinot-connectors/pinot-spark-3-connector/documentation/read_model.md [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+```scala
+val df = spark.read
+      .format("pinot")
+      .option("table", "airlineStats")
+      .option("tableType", "offline")
+      .load()
+      .filter($"DestStateName" === "Florida")
+      .filter($"Origin" === "ORD")
+      .select($"Carrier")
```
Comment (GSharayu) [inline pinot-connectors/pinot-spark-3-connector/documentation/read_model.md:123]: minor, there should be all 3 columns?

File: pinot-connectors/pinot-spark-3-connector/pom.xml (hunk 1 above)
Comment (GSharayu) [reply pinot-connectors/pinot-spark-3-connector/pom.xml:153]: +1

File: pinot-connectors/pinot-spark-3-connector/src/main/scala/org/apache/pinot/connector/spark/v3/datasource/PinotDataSource.scala [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+import org.apache.pinot.connector.spark.common.{PinotClusterClient, PinotDataSourceReadOptions}
+import org.apache.spark.sql.connector.catalog.{Table, TableProvider}
+import org.apache.spark.sql.connector.expressions.Transform
+import org.apache.spark.sql.sources.DataSourceRegister
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.util.CaseInsensitiveStringMap
+
+import java.util
+
+class PinotDataSource extends TableProvider with DataSourceRegister {
```
Comment (GSharayu) [inline pinot-connectors/pinot-spark-3-connector/src/main/scala/org/apache/pinot/connector/spark/v3/datasource/PinotDataSource.scala:30]: Can we please have javadocs over classes wherever important and possible, it helps in ease of understanding

File: pinot-connectors/pinot-spark-3-connector/README.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    software distributed under the License is distributed on an
+    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+    KIND, either express or implied.  See the License for the
+    specific language governing permissions and limitations
+    under the License.
+
+-->
+# Spark-Pinot Connector
+
+Spark-pinot connector to read and write data from/to Pinot.
```
Comment (jackjlli) [inline pinot-connectors/pinot-spark-3-connector/README.md:23]: Is this PR just related to read data not to write, right? If that's the case, can we state it here?

File: pinot-connectors/pinot-spark-3-connector/src/test/resources/schema/pinot-schema.json [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+  },
+  "dateTimeFieldSpecs": [
+    {
+      "name": "timestampField",
+      "dataType": "TIMESTAMP",
+      "format": "1:MILLISECONDS:EPOCH",
+      "granularity": "1:SECONDS"
+    }
+  ]
+}
```
Comment (jackjlli) [inline pinot-connectors/pinot-spark-3-connector/src/test/resources/schema/pinot-schema.json:75]: Missing empty line at the bottom.

File: pinot-connectors/pinot-spark-3-connector/src/test/resources/schema/spark-schema.json [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    "type" : "integer",
+    "nullable" : true,
+    "metadata" : { }
+  }, {
+    "name" : "timestampField",
+    "type" : "long",
+    "nullable" : true,
+    "metadata" : { }
+  } ]
+}
```
Comment (jackjlli) [inline pinot-connectors/pinot-spark-3-connector/src/test/resources/schema/spark-schema.json:105]: Same here.

File: pinot-connectors/pinot-spark-3-connector/pom.xml (hunk 1 above)
Comment (cbalci) [reply pinot-connectors/pinot-spark-3-connector/pom.xml:153]: Thanks for the suggestion. I checked the [dependency tree](https://gist.github.com/cbalci/f7b09e2df40f4c6731653584d9e51026) and it looks like the only hadoop dependencies are included in 'provided' scope, so I don't think we need to shade them here.
Let me know if you think otherwise.

File: pinot-connectors/pinot-spark-3-connector/documentation/read_model.md [hunk 8]
```diff
@@ -0,0 +1,140 @@
+<!--
```
Comment (cbalci) [inline pinot-connectors/pinot-spark-3-connector/documentation/read_model.md:1]: Full Disclosure: I copied this documentation from the Spark2 based implementation since at this point the functionality is pretty much same. So it is possible that I overlooked at some details, but always happy to fix.

File: pinot-connectors/pinot-spark-3-connector/README.md [hunk 9]
```diff
@@ -0,0 +1,69 @@
+<!--
```
Comment (cbalci) [inline pinot-connectors/pinot-spark-3-connector/README.md:1]: Disclosure: I copied this documentation from the Spark2 based implementation since at this point the functionality is pretty much same. However it is likely they will diverge pretty soon so I think they deserve separate doc pages.

File: pinot-connectors/pinot-spark-3-connector/src/main/scala/org/apache/pinot/connector/spark/v3/datasource/PinotDataSource.scala (hunk 4 above)
Comment (cbalci) [reply pinot-connectors/pinot-spark-3-connector/src/main/scala/org/apache/pinot/connector/spark/v3/datasource/PinotDataSource.scala:34]: Thanks for the suggestion. Added.

File: pinot-connectors/pinot-spark-3-connector/documentation/read_model.md (hunk 3 above)
Comment (cbalci) [reply pinot-connectors/pinot-spark-3-connector/documentation/read_model.md]: Good catch, fixed.

File: pinot-connectors/pinot-spark-3-connector/src/test/resources/schema/spark-schema.json (hunk 7 above)
Comment (cbalci) [reply pinot-connectors/pinot-spark-3-connector/src/test/resources/schema/spark-schema.json]: Added.

**Comment by cbalci** [comment]
from @KKcorps :
> Can you try to test it in cluster mode while running on a proper YARN, AWS EMR or DataProc cluster.

Good suggestion. I ended up testing this in our YARN environment with success. Although, I have to note that our YARN environment runs Spark 3.0.2 so I had to downgrade Spark version of the connector for the test. Everything else worked as expected.

**Comment by cbalci** [comment]
Thanks for the reviews @KKcorps @GSharayu @jackjlli ! I tried to address/resolve your comments, ptal.

**Review by KKcorps — APPROVED** [review]
LGTM! THanks for testing it out on YARN. 
One suggestion would be to add a bash command to trigger this job in cluster mode that you use on YARN cluster.

Sometimes running the same command as local but adding `deploy-mode cluster` doesn't work properly.

**Comment by GSharayu** [comment]
LGTM as well. Thank you for the changes

**Comment by cbalci** [comment]
Thanks for the reviews folks!

> One suggestion would be to add a bash command to trigger this job in cluster mode that you use on YARN cluster.
> 
> Sometimes running the same command as local but adding `deploy-mode cluster` doesn't work properly.

Good idea. Added a sample `spark-submit` command to the documentation for running the included examples in cluster mode.

**Comment by xiangfu0** [comment]
LGTM! Thanks for your contribution @cbalci !
